msgraph/core/middleware/retry.py

```diff
@@ -1,221 +1,48 @@
-
-        status_codes: [int] = kwargs.pop('retry_on_status_codes', [])
-
-        self._retry_on_status_codes: set = set(status_codes) | self._DEFAULT_RETRY_STATUS_CODES
-        self._allowed_methods: set = frozenset(
-            ['HEAD', 'GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS']
-        )
-        self._respect_retry_after_header: bool = True
-
-    @classmethod
-    def disable_retries(cls):
-        """
-        Disable retries by setting retry_total to zero.
-        retry_total takes precedence over all other counts.
-        """
-        return cls(max_retries=0)
-
-    def get_retry_options(self, middleware_control):
-        """
-        Check if request specific configs have been passed and override any session defaults
-        Then configure retry settings into the form of a dict.
-        """
-        if middleware_control:
-            return {
-                'total':
-                min(middleware_control.get('max_retries', self.max_retries), self.MAX_RETRIES),
-                'backoff':
-                middleware_control.get('retry_backoff_factor', self.backoff_factor),
-                'max_backoff':
-                middleware_control.get('retry_backoff_max', self.backoff_max),
-                'timeout':
-                middleware_control.get('retry_time_limit', self.timeout),
-                'retry_codes':
-                set(middleware_control.get('retry_on_status_codes', self._retry_on_status_codes))
-                | set(self._DEFAULT_RETRY_STATUS_CODES),
-                'methods':
-                self._allowed_methods,
-            }
-        return {
-            'total': self.max_retries,
-            'backoff': self.backoff_factor,
-            'max_backoff': self.backoff_max,
-            'timeout': self.timeout,
-            'retry_codes': self._retry_on_status_codes,
-            'methods': self._allowed_methods,
-        }
-
-    def send(self, request, **kwargs):
+    async def send(self, request: GraphRequest, transport: httpx.AsyncBaseTransport):
         """
         Sends the http request object to the next middleware or retries the request if necessary.
         """
-        retry_options = self.get_retry_options(request.context.middleware_control)
-        absolute_time_limit = min(retry_options['timeout'], self.MAX_DELAY)
         response = None
         retry_count = 0
-        retry_valid = True
+        retry_valid = self.retries_allowed
+
+        request.context.feature_usage = FeatureUsageFlag.RETRY_HANDLER_ENABLED
 
         while retry_valid:
             start_time = time.time()
             if retry_count > 0:
-                request.headers.update({'retry-attempt': '{}'.format(retry_count)})
-            response = super().send(request, **kwargs)
+                request.headers.update({'retry-attempt': f'{retry_count}'})
+            response = await super(RetryHandler, self).send(request, transport)
```

Comment:
Since we have overridden the send method of the retry handler in http core with a custom one in `GraphRetryHandler`, `super(RetryHandler, self).send` calls the send method of the parent class of all middleware which is `BaseMiddleware`. Therefore there is no double nesting.

Reply:
Ah. Thanks for that!